### PR TITLE
MF-196 cohert visibility levy

### PR DESCRIPTION
**Pull request by ChughAkshay**

### CODE REVIEW

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenValidatingCreateReservationLevyEmployerCommand.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+using Castle.Core.Internal;
+using NUnit.Framework;
+using SFA.DAS.Reservations.Application.Reservations.Commands.CreateReservationLevyEmployer;
+
+namespace SFA.DAS.Reservations.Application.UnitTests.Reservations.Commands.CreateReservationLevyEmployer
+{
+    [TestFixture]
+    public class WhenValidatingCreateReservationLevyEmployerCommand
+    {
+        [TestCase(Int64.MinValue, Int64.MinValue)]
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenValidatingCreateReservationLevyEmployerCommand.cs]: No sure we need to test this boundaries, another test that should be done is to make sure that if one of the two required values is valid and the other is not then the result is not valid

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenValidatingCreateReservationLevyEmployerCommand.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            var result = await validator.ValidateAsync(command);
+
+            //Assert
+            Assert.False(result.IsValid());
+            Assert.True(result.ValidationDictionary.ContainsKey(nameof(command.AccountId)));
+            Assert.True(result.ValidationDictionary.ContainsKey(nameof(command.AccountLegalEntityId)));
+
+        }
+
+        [TestCase(Int64.MaxValue, Int64.MaxValue)]
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenValidatingCreateReservationLevyEmployerCommand.cs]: Dont think we need to do these boundary tests

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenCreatingANewReservationForLevyEmployer.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+            service.Setup(x =>
+                    x.CreateReservationLevyEmployer(It.IsAny<Guid>(), request.AccountId, request.AccountLegalEntityId))
+                .ReturnsAsync(new CreateReservationResponse(){Id = id});
+
+            //Act
+            var result = await handler.Handle(request, CancellationToken.None);
+
+            //Assert
+            Assert.AreEqual(id,result.ReservationId);
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenCreatingANewReservationForLevyEmployer.cs]: The assertion of this test can be moved up to the other test as the behaviour is covered there

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenCreatingANewReservationForLevyEmployer.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        public void ThenTheCommandGetsValidatedAndThrowsAnExceptionIfNotValid(
+            CreateReservationLevyEmployerCommand request,
+            [Frozen] Mock<IValidator<CreateReservationLevyEmployerCommand>> validator,
+            CreateReservationLevyEmployerCommandHandler handler)
+        {
+            //Arrange
+            validator.Setup(x => x.ValidateAsync(It.IsAny<CreateReservationLevyEmployerCommand>()))
+                .ReturnsAsync(new ValidationResult {ValidationDictionary = new Dictionary<string, string> {{"", ""}}});
+            //Act
+            Assert.ThrowsAsync<ValidationException>(async() => await handler.Handle(request, CancellationToken.None));
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenCreatingANewReservationForLevyEmployer.cs]: `async` and `await` can be removed

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Queries/GetAccountReservationStatus/WhenIValidateGetAccountReservationStatusQuery.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        public async Task AndTheQueryIsValid_ThenValidationPasses()
+        {
+            //Arrange
+            var query = new GetAccountReservationStatusQuery(){AccountId = 12342};
+
+            //Act
+            var result = await _validator.ValidateAsync(query);
+
+            //Assert
+            Assert.AreEqual(true,result.IsValid());
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Queries/GetAccountReservationStatus/WhenIValidateGetAccountReservationStatusQuery.cs]:
```suggestion
            Assert.IsTrue(result.IsValid());
```

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Queries/GetAccountReservationStatus/WhenIValidateGetAccountReservationStatusQuery.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        public async Task AndTheAccountIdIsDefault_ThenValidationFails()
+        {
+            //Arrange
+            var query = new GetAccountReservationStatusQuery() { AccountId = default(long) };
+
+            //Act
+            var result = await _validator.ValidateAsync(query);
+
+            //Assert
+            Assert.AreEqual(false, result.IsValid());
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Queries/GetAccountReservationStatus/WhenIValidateGetAccountReservationStatusQuery.cs]:
```suggestion
            Assert.IsFalse(result.IsValid());
```

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenValidatingCreateReservationLevyEmployerCommand.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            var validator = new CreateReservationLevyEmployerCommandValidator();
+            command.AccountId = accountId;
+            command.AccountLegalEntityId = accountLegalEntityId;
+
+            //Act
+            var result = await validator.ValidateAsync(command);
+
+            //Assert
+            Assert.True(result.IsValid());
+            Assert.True(result.ValidationDictionary.IsNullOrEmpty());
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenValidatingCreateReservationLevyEmployerCommand.cs]: Don't need to assert on this

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenCreatingANewReservationForLevyEmployer.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            //Arrange
+            validator.Setup(x => x.ValidateAsync(request))
+                .ReturnsAsync(new ValidationResult {ValidationDictionary = new Dictionary<string, string>()});
+
+            service.Setup(x =>
+                    x.CreateReservationLevyEmployer(It.IsAny<Guid>(), request.AccountId, request.AccountLegalEntityId))
+                .ReturnsAsync(new CreateReservationResponse());
+
+            //Act
+            var result = await handler.Handle(request, CancellationToken.None);
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/CreateReservationLevyEmployer/WhenCreatingANewReservationForLevyEmployer.cs]: `result` isn't used and can be removed.

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Services/WhenCallingReservationService.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+                .ReturnsAsync(new CreateReservationResponse() {Id = id});
+            //Act
+            await service.CreateReservationLevyEmployer(id,accountId,accountLegalEntityId);
+
+            //Assert
+            mockApiClient.Verify(x => x.Create<CreateReservationResponse>(It.Is<ReservationApiRequest>(
+                request => request.Id == id &&
+                           request.AccountId == accountId &&
+                           request.AccountLegalEntityId == accountLegalEntityId &&
+                           request.IsLevyAccount == true)));
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Services/WhenCallingReservationService.cs]: `==true` can be removed on the IsLevyAccount

File: src/SFA.DAS.Reservations.Application/Reservations/Queries/GetAccountReservationStatus/GetAccountReservationStatusQueryHandler.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            {
+                throw new ValidationException(
+                    new ValidationResult("The following parameters have failed validation", validationResult.ErrorList), null, null);
+            }
+
+            var response =
+                await _apiClient.Get<AccountReservationStatusResponse>(
+                    new AccountReservationStatusRequest(_config.Url, request.AccountId));
+
+            return new GetAccountReservationStatusResponse(){CanAutoCreateReservations = response.CanAutoCreateReservations};
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application/Reservations/Queries/GetAccountReservationStatus/GetAccountReservationStatusQueryHandler.cs]:
```suggestion
            return new GetAccountReservationStatusResponse{CanAutoCreateReservations = response.CanAutoCreateReservations};
```

File: src/SFA.DAS.Reservations.Domain/Reservations/Api/AccountReservationStatusRequest.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        public AccountReservationStatusRequest(string baseUrl, long accountId)
+        {
+            BaseUrl = baseUrl;
+            AccountId = accountId;
+        }
+        
+        public long AccountId { get; }
+        public string BaseUrl { get; }
+        public string GetUrl => $"{BaseUrl}api/accounts/{AccountId}/status";
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Domain/Reservations/Api/AccountReservationStatusRequest.cs]: Not sure if we have a test around this URL being correctly set that I could see in this PR

File: src/SFA.DAS.Reservations.Web.UnitTests/Reservations/WhenCallingGetSelectReservation.cs [hunk 12]
```diff
@@ -175,5 +182,143 @@ public class WhenCallingGetSelectReservation
...
+                .ReturnsAsync(accountStatusResponse);
+
+
+            //Act
+            var result = await controller.SelectReservation(routeModel, viewModel);
+
+
+            //Assert
+            _mediator.Verify(x =>
+                x.Send(It.IsAny<GetAccountReservationStatusQuery>(),CancellationToken.None),Times.Once());
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Web.UnitTests/Reservations/WhenCallingGetSelectReservation.cs]: Can we change this to make sure the correct params are going into the query

**Comment by ChughAkshay** [comment]
Added code review changes
